frontend/src/metabase/browse/utils.ts

```diff
@@ -41,16 +47,18 @@ export type ModelFilterControlsProps = {
  * or false if it is inactive */
 export type ActualModelFilters = Record<string, boolean>;
 
-export const filterModels = (
-  unfilteredModels: ModelResult[],
+export const filterModels = <T extends ModelResult[] | RecentCollectionItem[]>(
+  unfilteredModels: T,
   actualModelFilters: ActualModelFilters,
   availableModelFilters: AvailableModelFilters,
-) => {
+): Array<T[number]> => {
   return _.reduce(
     actualModelFilters,
     (acc, shouldFilterBeActive, filterName) =>
       shouldFilterBeActive
-        ? acc.filter(availableModelFilters[filterName].predicate)
+        ? // https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#easier-method-usage-for-unions-of-arrays
+          // @ts-expect-error This would be valid without a cast in TS 5.2+
+          (acc.filter(availableModelFilters[filterName].predicate) as T)
```

Comment:
Thanks @chriskrycho for help with this voodoo 🙏 

see: https://github.com/metabase/metabase/issues/42704

Reply:
Fuller credit should also include @dfreeman, who helped *me* figure it out! Fist bumps all around. 😅